v2/cmd/driver/main.go

```diff
@@ -0,0 +1,176 @@
+// Copyright 2021 The Kubeflow Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/golang/glog"
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/kubeflow/pipelines/api/v2alpha1/go/pipelinespec"
+	"github.com/kubeflow/pipelines/v2/driver"
+	"github.com/kubeflow/pipelines/v2/metadata"
+)
+
+const (
+	driverTypeArg = "type"
+)
+
+var (
+	// inputs
+	driverType        = flag.String(driverTypeArg, "", "task driver type, one of ROOT_DAG, CONTAINER")
+	pipelineName      = flag.String("pipeline_name", "", "pipeline context name")
+	runID             = flag.String("run_id", "", "pipeline run uid")
+	componentSpecJson = flag.String("component", "{}", "component spec")
+	taskSpecJson      = flag.String("task", "{}", "task spec")
+	runtimeConfigJson = flag.String("runtime_config", "{}", "jobruntime config")
+
+	// container inputs
+	dagContextID   = flag.Int64("dag_context_id", 0, "DAG context ID")
+	dagExecutionID = flag.Int64("dag_execution_id", 0, "DAG execution ID")
+
+	// config
+	mlmdServerAddress = flag.String("mlmd_server_address", "", "MLMD server address")
+	mlmdServerPort    = flag.String("mlmd_server_port", "", "MLMD server port")
+
+	// output paths
+	executionIDPath   = flag.String("execution_id_path", "", "Exeucution ID output path")
+	contextIDPath     = flag.String("context_id_path", "", "Context ID output path")
+	executorInputPath = flag.String("executor_input_path", "", "Executor Input output path")
+)
+
+// func RootDAG(pipelineName string, runID string, component *pipelinespec.ComponentSpec, task *pipelinespec.PipelineTaskSpec, mlmd *metadata.Client) (*Execution, error) {
+
+func main() {
+	flag.Parse()
+	err := validate()
+	if err != nil {
+		glog.Exitf("%v", err)
```

Comment:
Do we need to add prefix here like "Failed to validate xxx"

Reply:
That's a good catch! A small difference is that I would change the wording to "failed to run KFP driver:", because consider that we have many functions, some functions are called once, some functions are called multiple times.
If we add error message to explain what failed at call site, we'll have to repeat for each function call.
Instead, so far I prefer adding error message to explain what failed inside the function itself.
This way, there's better locality and also I only need to add the error message once per function.
Therefore, because this is the driver CLI function, I'll draft error message related to the driver and let the function I call describe their own error messages.

Reply:
Updated